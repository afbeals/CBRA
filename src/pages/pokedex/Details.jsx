// Extenal
import React, { useEffect } from 'react';
import PropTypes from 'prop-types';

// Local
import withPokeList from '~Modules/pokelist/withPokeList';
import { ErrorBoundary } from '~Components';
import Image from './details/Image';
import Title from './details/Title';
import BaseStats from './details/BaseStats';
import StatsDisplay from './details/StatsDisplay';
import Abilities from './details/Abilities';

// Constants
const classname = 'details';

// Component
const Details = ({
  pClassname,
  fetchPokemon,
  fetchPokemonCancel,
  getIsFetchingPokemon: isFetching,
  getSelectedPokemon: pokemon,
}) => {
  const sprites = pokemon?.sprites ?? null;
  const name = pokemon?.name ?? null;
  const height = pokemon?.height ?? null;
  const weight = pokemon?.weight ?? null;
  const types = pokemon?.types ?? [];
  const pokemonId = pokemon?.id ?? null;
  const stats = pokemon?.stats ?? [];
  const moves = pokemon?.moves ?? [];
  const abilities = pokemon?.abilities ?? [];

  useEffect(() => {
    if (pokemon && !pokemon.height) {
      fetchPokemon();
    }
  }, [pokemon, fetchPokemon]);

  useEffect(
    () => () => {
      if (isFetching) {
        fetchPokemonCancel();
      }
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [],
  );

  return (
    <ErrorBoundary>
      <div data-testid="details" className={`${pClassname}__${classname}`}>
        {/* {isFetching && <Loading text="pokemon" />} */}
        <Title
          pClassname={`${pClassname}__${classname}`}
          name={name}
          id={pokemonId}
        />
        <Image
          pClassname={`${pClassname}__${classname}`}
          sprites={sprites}
          name={name}
        />
        <BaseStats
          pClassname={`${pClassname}__${classname}`}
          height={height}
          weight={weight}
          types={types}
        />
        <Abilities
          pClassname={`${pClassname}__${classname}`}
          moves={moves}
          abilities={abilities}
        />
        <StatsDisplay
          pClassname={`${pClassname}__${classname}`}
          stats={stats}
        />
      </div>
    </ErrorBoundary>
  );
};

Details.propTypes = {
  pClassname: PropTypes.string.isRequired,
  getIsFetchingPokemon: PropTypes.bool.isRequired,
  fetchPokemonCancel: PropTypes.func.isRequired,
  fetchPokemon: PropTypes.func.isRequired,
  getSelectedPokemon: PropTypes.object,
};

export default withPokeList(Details);

export { Details };
