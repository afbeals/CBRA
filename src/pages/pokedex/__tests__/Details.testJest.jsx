// External
import React from 'react';
import { render, cleanup } from '@testing-library/react';

// Local
import { Details } from '../Details';

// tests
describe('Details', () => {
  afterEach(cleanup);
  it('should match Details to snapshot', () => {
    const jestFn = jest.fn();
    const { getByTestId } = render(
      <Details
        pClassname="parent"
        getIsFetchingPokemon={false}
        fetchPokemonCancel={jestFn}
        fetchPokemon={jestFn}
        getSelectedPokemon={{}}
      />,
    );
    const details = getByTestId('details');
    expect(details).toMatchSnapshot();
  });
});
