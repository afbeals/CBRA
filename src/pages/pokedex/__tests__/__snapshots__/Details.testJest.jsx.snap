// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details should match Details to snapshot 1`] = `
<div
  class="parent__details"
  data-testid="details"
>
  <div
    class="parent__details__title"
    data-testid="details-title"
  >
    <h3
      class="parent__details__title__name"
    />
    <h3
      class="parent__details__title__id"
    />
  </div>
  <div
    class="parent__details__image"
    data-testid="details-image"
  >
    <div
      class="parent__details__image__gender"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root male iconButton MuiIconButton-colorSecondary Mui-disabled MuiIconButton-sizeSmall Mui-disabled"
        data-testid="iconButton"
        disabled=""
        role="button"
        tabindex="-1"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <i
            class="fas fa-mars"
          />
        </span>
      </button>
      <button
        class="MuiButtonBase-root MuiIconButton-root female iconButton MuiIconButton-colorPrimary Mui-disabled MuiIconButton-sizeSmall Mui-disabled"
        data-testid="iconButton"
        disabled=""
        role="button"
        tabindex="-1"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <i
            class="fas fa-venus"
          />
        </span>
      </button>
    </div>
    <div
      class="parent__details__image__flipper"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root iconButton MuiIconButton-colorPrimary Mui-disabled MuiIconButton-sizeSmall Mui-disabled"
        data-testid="iconButton"
        disabled=""
        role="button"
        tabindex="-1"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <i
            class="fas fa-sync-alt"
          />
        </span>
      </button>
    </div>
  </div>
  <div
    class="parent__details__base"
    data-testid="details-baseStats"
  >
    <div
      class="parent__details__base__info"
    >
      <div
        class="section section--info"
      >
        <span
          class="section__title"
        >
          height:
        </span>
        <p
          class="section__content"
        />
      </div>
      <div
        class="section section--info"
      >
        <span
          class="section__title"
        >
          weight:
        </span>
        <p
          class="section__content"
        />
      </div>
      <div
        class="section section--types"
      />
    </div>
  </div>
  <div
    class="parent__details__abilities"
    data-testid="details-abilites"
  >
    <div
      class="section"
    >
      <p
        class="section__title"
      >
        abilities
      </p>
      <ul
        class="section__list"
      />
    </div>
  </div>
  <div
    class="parent__details__stats"
    data-testid="details-stats"
  >
    <ul
      class="parent__details__stats__list"
    />
  </div>
</div>
`;
